File: app/services/github_import_service.py
"""Односторонняя синхронизация GitHub issues -> задачи.

Осознанное отступление от workflow (см. docs/decisions.md): закрытый на GitHub
issue переводит задачу в done напрямую, без review/approve и без начисления XP —
awarder здесь не вызывается, ExperienceLog не затрагивается. Переоткрытие issue
статус задачи не понижает.
"""

import secrets

from app.core.security import hash_password
from app.integrations.github_client import GitHubIssue, GitHubIssueSource
from app.models.enums import TaskStatus
from app.models.task import Task
from app.repositories.task_repo import TaskRepository
from app.repositories.user_repo import UserRepository

_TITLE_MAX_LENGTH = 200


class GitHubImportService:
    def __init__(
        self,
        source: GitHubIssueSource,
        task_repo: TaskRepository,
        user_repo: UserRepository,
    ) -> None:
        self.source = source
        self.task_repo = task_repo
        self.user_repo = user_repo

    async def sync(self, current_user_id: int) -> dict:
        issues = await self.source.fetch_issues()

        created = 0
        updated = 0
        users_created = 0
        assignee_cache: dict[str, int] = {}
        label_cache: dict[str, int] = {}

        for issue in issues:
            assignee_id: int | None = None
            if issue.assignee_login is not None:
                assignee_id, was_created = await self._resolve_assignee(
                    issue.assignee_login, assignee_cache
                )
                if was_created:
                    users_created += 1

            label_ids = [await self._resolve_label(name, label_cache) for name in issue.labels]

            task = await self.task_repo.get_task_by_github_issue_number(issue.number)
            if task is None:
                await self._create_task(issue, assignee_id, label_ids, current_user_id)
                created += 1
            elif await self._update_task(task, issue, assignee_id, label_ids):
                updated += 1

        return {"created": created, "updated": updated, "users_created": users_created}

    async def _create_task(
        self,
        issue: GitHubIssue,
        assignee_id: int | None,
        label_ids: list[int],
        creator_id: int,
    ) -> None:
        task = await self.task_repo.create_task(
            {
                "title": issue.title[:_TITLE_MAX_LENGTH],
                "description": issue.body,
                "status": self._map_status(issue.state),
                "assignee_id": assignee_id,
                "github_issue_number": issue.number,
            },
            creator_id=creator_id,
        )
        if label_ids:
            await self.task_repo.replace_task_labels(task.id, label_ids)

    async def _update_task(
        self,
        task: Task,
        issue: GitHubIssue,
        assignee_id: int | None,
        label_ids: list[int],
    ) -> bool:
        fields: dict = {}

        title = issue.title[:_TITLE_MAX_LENGTH]
        if task.title != title:
            fields["title"] = title
        if task.description != issue.body:
            fields["description"] = issue.body

        # GitHub может только закрывать: переоткрытие статус не понижает.
        if issue.state == "closed" and task.status != TaskStatus.done:
            fields["status"] = TaskStatus.done

        # Снятого на GitHub исполнителя не трогаем — не затираем локальные назначения.
        if assignee_id is not None and task.assignee_id != assignee_id:
            fields["assignee_id"] = assignee_id

        changed = False
        if fields:
            await self.task_repo.update_task(task, fields)
            changed = True

        current_label_ids = {task_label.label_id for task_label in task.labels}
        if current_label_ids != set(label_ids):
            await self.task_repo.replace_task_labels(task.id, label_ids)
            changed = True

        return changed

    @staticmethod
    def _map_status(state: str) -> TaskStatus:
        return TaskStatus.done if state == "closed" else TaskStatus.todo

    async def _resolve_assignee(self, login: str, cache: dict[str, int]) -> tuple[int, bool]:
        if login in cache:
            return cache[login], False

        user = await self.user_repo.get_user_by_github_login(login)
        if user is not None:
            cache[login] = user.id
            return user.id, False

        username = login
        suffix = 1
        while await self._identity_taken(username):
            suffix += 1
            username = f"{login}-{suffix}"

        user = await self.user_repo.create_placeholder_user(
            github_login=login,
            # Реальная noreply-конвенция GitHub; вход таким профилем невозможен.
            email=f"{username}@users.noreply.github.com",
            username=username,
            hashed_password=hash_password(secrets.token_urlsafe(32)),
        )
        cache[login] = user.id
        return user.id, True

    async def _identity_taken(self, username: str) -> bool:
        if await self.user_repo.get_user_by_username(username) is not None:
            return True
        email = f"{username}@users.noreply.github.com"
        return await self.user_repo.get_user_by_email(email) is not None

    async def _resolve_label(self, name: str, cache: dict[str, int]) -> int:
        key = name.lower()
        if key in cache:
            return cache[key]

        label = await self.task_repo.get_label_by_name_ci(name)
        if label is None:
            label = await self.task_repo.create_label(name, None)
        cache[key] = label.id
        return label.id

File: backend/src/main/java/com/skilltracker/service/GitHubImportService.java
package com.skilltracker.service;

import com.skilltracker.domain.Label;
import com.skilltracker.domain.Task;
import com.skilltracker.domain.TaskLabel;
import com.skilltracker.domain.TaskStatus;
import com.skilltracker.domain.User;
import com.skilltracker.dto.GitHubSyncResponse;
import com.skilltracker.integration.GitHubIssue;
import com.skilltracker.integration.GitHubIssueSource;
import com.skilltracker.repository.LabelRepository;
import com.skilltracker.repository.TaskLabelRepository;
import com.skilltracker.repository.TaskRepository;
import com.skilltracker.repository.UserRepository;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * One-way synchronisation of GitHub issues into tasks.
 *
 * <p>A deliberate departure from the normal workflow (see docs/decisions.md): an issue closed on
 * GitHub moves its task straight to {@code done} without review, approval or an XP award, and
 * reopening an issue never demotes the task.
 */
@Service
public class GitHubImportService {

    private static final int TITLE_MAX_LENGTH = 200;
    private static final String NOREPLY_DOMAIN = "@users.noreply.github.com";

    private final GitHubIssueSource source;
    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final LabelRepository labelRepository;
    private final TaskLabelRepository taskLabelRepository;
    private final PasswordEncoder passwordEncoder;
    private final SecureRandom secureRandom = new SecureRandom();

    public GitHubImportService(
            GitHubIssueSource source,
            TaskRepository taskRepository,
            UserRepository userRepository,
            LabelRepository labelRepository,
            TaskLabelRepository taskLabelRepository,
            PasswordEncoder passwordEncoder) {
        this.source = source;
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
        this.labelRepository = labelRepository;
        this.taskLabelRepository = taskLabelRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Transactional
    public GitHubSyncResponse sync(Integer currentUserId) {
        List<GitHubIssue> issues = source.fetchIssues();

        int created = 0;
        int updated = 0;
        int usersCreated = 0;
        Map<String, Integer> assigneeCache = new HashMap<>();
        Map<String, Integer> labelCache = new HashMap<>();

        for (GitHubIssue issue : issues) {
            Integer assigneeId = null;
            if (issue.assigneeLogin() != null) {
                ResolvedUser resolved = resolveAssignee(issue.assigneeLogin(), assigneeCache);
                assigneeId = resolved.userId();
                if (resolved.wasCreated()) {
                    usersCreated++;
                }
            }

            List<Integer> labelIds = issue.labels().stream()
                    .map(name -> resolveLabel(name, labelCache))
                    .toList();

            Optional<Task> existing = taskRepository.findByGithubIssueNumber(issue.number());
            if (existing.isEmpty()) {
                createTask(issue, assigneeId, labelIds, currentUserId);
                created++;
            } else if (updateTask(existing.get(), issue, assigneeId, labelIds)) {
                updated++;
            }
        }

        return new GitHubSyncResponse(created, updated, usersCreated);
    }

    private void createTask(GitHubIssue issue, Integer assigneeId, List<Integer> labelIds, Integer creatorId) {
        Task task = new Task(truncateTitle(issue.title()), creatorId);
        task.setDescription(issue.body());
        task.setStatus(mapStatus(issue.state()));
        task.setAssigneeId(assigneeId);
        task.setGithubIssueNumber(issue.number());
        taskRepository.saveAndFlush(task);

        if (!labelIds.isEmpty()) {
            replaceLabels(task.getId(), labelIds);
        }
    }

    private boolean updateTask(Task task, GitHubIssue issue, Integer assigneeId, List<Integer> labelIds) {
        boolean changed = false;

        String title = truncateTitle(issue.title());
        if (!title.equals(task.getTitle())) {
            task.setTitle(title);
            changed = true;
        }
        if (!java.util.Objects.equals(task.getDescription(), issue.body())) {
            task.setDescription(issue.body());
            changed = true;
        }
        // GitHub can only close a task here: reopening an issue never demotes a finished task.
        if ("closed".equals(issue.state()) && task.getStatus() != TaskStatus.DONE) {
            task.setStatus(TaskStatus.DONE);
            changed = true;
        }
        // An assignee cleared on GitHub is left alone so local assignments are not wiped.
        if (assigneeId != null && !assigneeId.equals(task.getAssigneeId())) {
            task.setAssigneeId(assigneeId);
            changed = true;
        }

        if (changed) {
            taskRepository.saveAndFlush(task);
        }

        Set<Integer> currentLabelIds = taskLabelRepository.findByTaskIdOrderById(task.getId()).stream()
                .map(TaskLabel::getLabelId)
                .collect(Collectors.toSet());
        if (!currentLabelIds.equals(new HashSet<>(labelIds))) {
            replaceLabels(task.getId(), labelIds);
            changed = true;
        }

        return changed;
    }

    private void replaceLabels(Integer taskId, List<Integer> labelIds) {
        taskLabelRepository.deleteAllByTaskId(taskId);
        labelIds.forEach(labelId -> taskLabelRepository.save(new TaskLabel(taskId, labelId)));
        taskLabelRepository.flush();
    }

    private record ResolvedUser(Integer userId, boolean wasCreated) {}

    private ResolvedUser resolveAssignee(String login, Map<String, Integer> cache) {
        Integer cached = cache.get(login);
        if (cached != null) {
            return new ResolvedUser(cached, false);
        }

        Optional<User> existing = userRepository.findByGithubLogin(login);
        if (existing.isPresent()) {
            cache.put(login, existing.get().getId());
            return new ResolvedUser(existing.get().getId(), false);
        }

        String username = login;
        int suffix = 1;
        while (isIdentityTaken(username)) {
            suffix++;
            username = login + "-" + suffix;
        }

        User placeholder = new User(
                // The real GitHub noreply convention; signing in with such a profile is impossible.
                username + NOREPLY_DOMAIN, username, passwordEncoder.encode(randomSecret()), "user");
        placeholder.setGithubLogin(login);
        placeholder.setPlaceholder(true);
        userRepository.saveAndFlush(placeholder);

        cache.put(login, placeholder.getId());
        return new ResolvedUser(placeholder.getId(), true);
    }

    private boolean isIdentityTaken(String username) {
        return userRepository.existsByUsername(username) || userRepository.existsByEmail(username + NOREPLY_DOMAIN);
    }

    private Integer resolveLabel(String name, Map<String, Integer> cache) {
        String key = name.toLowerCase(Locale.ROOT);
        Integer cached = cache.get(key);
        if (cached != null) {
            return cached;
        }

        Label label = labelRepository
                .findByNameIgnoreCase(name)
                .orElseGet(() -> labelRepository.saveAndFlush(new Label(name, null)));
        cache.put(key, label.getId());
        return label.getId();
    }

    private TaskStatus mapStatus(String state) {
        return "closed".equals(state) ? TaskStatus.DONE : TaskStatus.TODO;
    }

    private String truncateTitle(String title) {
        return title.length() > TITLE_MAX_LENGTH ? title.substring(0, TITLE_MAX_LENGTH) : title;
    }

    private String randomSecret() {
        byte[] bytes = new byte[24];
        secureRandom.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }
}
